[ao] Add line plot visualization capability to ModelReportVisualizer

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81975
* __->__ #81974
* #81973
* #81673

Summary: This adds the capability to visualize the line plot in the
ModelReportVisualizer. You can visualize line plots of single feature,
and this feature can either be a per-tensor or per-channel feature. If
the feature is per tensor, then the idx of the module is printed as the
x axis and the values of the feature as the y. If the feature is per
channel, then **an** (the first one) idx of the module will the value on
the x axis and the corresponding feature val in the y axis, and there
will be a seperate line for each channel, and a legend denoting which
line belongs to which channel.

Expected Usage
```
mod_rep_visualizer.generate_plot_visualization(<feature_name>)
```

You can also filter the modules so that only modules with a certain
substring will have their features represented in the plot.

> ** This is intended to be used in a `.ipynb` style notebook**

The tests for this were just visual inspection for two reasons:
    1.) This method does not return anything, it just generates the
    visualization plot
    2.) All the data to create the plot visualization is gotten from
    `generate_filtered_tables` which is already tested, so testing all that
    for this again would be redundant.

Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao] Add line plot visualization capability to ModelReportVisualizer

Summary: This adds the capability to visualize the line plot in the
ModelReportVisualizer. You can visualize line plots of single feature,
and this feature can either be a per-tensor or per-channel feature. If
the feature is per tensor, then the idx of the module is printed as the
x axis and the values of the feature as the y. If the feature is per
channel, then **an** (the first one) idx of the module will the value on
the x axis and the corresponding feature val in the y axis, and there
will be a seperate line for each channel, and a legend denoting which
line belongs to which channel.

Expected Usage
```
mod_rep_visualizer.generate_plot_visualization(<feature_name>)
```

You can also filter the modules so that only modules with a certain
substring will have their features represented in the plot.

> ** This is intended to be used in a `.ipynb` style notebook**

The tests for this were just visual inspection for two reasons:
    1.) This method does not return anything, it just generates the
    visualization plot
    2.) All the data to create the plot visualization is gotten from
    `generate_filtered_tables` which is already tested, so testing all that
    for this again would be redundant.

Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Add line plot visualization capability to ModelReportVisualizer"


Summary: This adds the capability to visualize the line plot in the
ModelReportVisualizer. You can visualize line plots of single feature,
and this feature can either be a per-tensor or per-channel feature. If
the feature is per tensor, then the idx of the module is printed as the
x axis and the values of the feature as the y. If the feature is per
channel, then **an** (the first one) idx of the module will the value on
the x axis and the corresponding feature val in the y axis, and there
will be a seperate line for each channel, and a legend denoting which
line belongs to which channel.

Expected Usage
```
mod_rep_visualizer.generate_plot_visualization(<feature_name>)
```

You can also filter the modules so that only modules with a certain
substring will have their features represented in the plot.

> ** This is intended to be used in a `.ipynb` style notebook**

The tests for this were just visual inspection for two reasons:
    1.) This method does not return anything, it just generates the
    visualization plot
    2.) All the data to create the plot visualization is gotten from
    `generate_filtered_tables` which is already tested, so testing all that
    for this again would be redundant.

Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
![thumbnail_IMG_1317](https://user-images.githubusercontent.com/35707033/180421027-c8b962f0-bbec-4dad-ab8c-cca7c29f398c.jpeg)
![thumbnail_IMG_1316](https://user-images.githubusercontent.com/35707033/180421030-3f51a520-99da-4959-9174-4d2519993df5.jpeg)

^^ Example of per channel plot and tensor plot

vspenubarthi committed:
Update on "[ao] Add line plot visualization capability to ModelReportVisualizer"


Summary: This adds the capability to visualize the line plot in the
ModelReportVisualizer. You can visualize line plots of single feature,
and this feature can either be a per-tensor or per-channel feature. If
the feature is per tensor, then the idx of the module is printed as the
x axis and the values of the feature as the y. If the feature is per
channel, then **an** (the first one) idx of the module will the value on
the x axis and the corresponding feature val in the y axis, and there
will be a seperate line for each channel, and a legend denoting which
line belongs to which channel.

Expected Usage
```
mod_rep_visualizer.generate_plot_visualization(<feature_name>)
```

You can also filter the modules so that only modules with a certain
substring will have their features represented in the plot.

> ** This is intended to be used in a `.ipynb` style notebook**

The tests for this were just visual inspection for two reasons:
    1.) This method does not return anything, it just generates the
    visualization plot
    2.) All the data to create the plot visualization is gotten from
    `generate_filtered_tables` which is already tested, so testing all that
    for this again would be redundant.

Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Add line plot visualization capability to ModelReportVisualizer"


Summary: This adds the capability to visualize the line plot in the
ModelReportVisualizer. You can visualize line plots of single feature,
and this feature can either be a per-tensor or per-channel feature. If
the feature is per tensor, then the idx of the module is printed as the
x axis and the values of the feature as the y. If the feature is per
channel, then **an** (the first one) idx of the module will the value on
the x axis and the corresponding feature val in the y axis, and there
will be a seperate line for each channel, and a legend denoting which
line belongs to which channel.

Expected Usage
```
mod_rep_visualizer.generate_plot_visualization(<feature_name>)
```

You can also filter the modules so that only modules with a certain
substring will have their features represented in the plot.

> ** This is intended to be used in a `.ipynb` style notebook**

The tests for this were just visual inspection for two reasons:
    1.) This method does not return anything, it just generates the
    visualization plot
    2.) All the data to create the plot visualization is gotten from
    `generate_filtered_tables` which is already tested, so testing all that
    for this again would be redundant.

Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -419,19 +426,105 @@ def generate_plot_visualization(self, feature_filter: str, module_fqn_filter: st
+            When the tensor information is plotted, it will plot:
+                idx as the x val, feature value as the y_val
+            When the channel information is plotted, it will plot:
+                idx // 3 as the x val, feature value as the y_val [for each channel]
```
what does `idx // 3` mean?

vspenubarthi replied:
updated with better comments

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -419,19 +426,105 @@ def generate_plot_visualization(self, feature_filter: str, module_fqn_filter: st
+        x_data: List = []
+        y_data: List[List] = []
+        # the feature will either be a tensor feature or channel feature
+        if tensor_info_features_count == 1:
```
can we just plot multiple figures if there are multiple selected features?

vspenubarthi replied:
since the plotting function is pretty straightforward, we just decided to just call separately for each thing they want to plot rather than build that functionality into the code itself

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -419,19 +426,105 @@ def generate_plot_visualization(self, feature_filter: str, module_fqn_filter: st
+            data_is_per_channel = True
+            # gather the x_data and multiple y_data
+            # calculate the number of channels
+            num_channels: int = max(row[2] for row in channel_table) + 1
```
what is `row[2]`?

vspenubarthi replied:
updated to be more descriptive

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -419,19 +426,105 @@ def generate_plot_visualization(self, feature_filter: str, module_fqn_filter: st
+            for row in tensor_table:
+                row_value = row[ModelReportVisualizer.DEFAULT_NON_FEATURE_TENSOR_HEADERS]
+                if not type(row_value) == str:
+                    x_data.append(row[0])
```
what is `row[0]`?

vspenubarthi replied:
refactored to not need

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -419,19 +426,105 @@ def generate_plot_visualization(self, feature_filter: str, module_fqn_filter: st
+                y_data.append([])  # seperate data list per channel
+
+            # keep track of last added x_value
+            last_added = None
```
what is this used for?

vspenubarthi replied:
refactored to not need

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -419,19 +426,105 @@ def generate_plot_visualization(self, feature_filter: str, module_fqn_filter: st
+        # keep track of per_channel or not
+        data_is_per_channel: bool = False
+
+        x_data: List = []
+        y_data: List[List] = []
```
I'm wondering if we can preprocess the data so that getting x_data and y_data are just simple dictionary lookup or selecting a subarray based on index

vspenubarthi replied:
changed so we are easily getting x and y data in a helper function

vspenubarthi committed:
Update on "[ao] Add line plot visualization capability to ModelReportVisualizer"


Summary: This adds the capability to visualize the line plot in the
ModelReportVisualizer. You can visualize line plots of single feature,
and this feature can either be a per-tensor or per-channel feature. If
the feature is per tensor, then the idx of the module is printed as the
x axis and the values of the feature as the y. If the feature is per
channel, then **an** (the first one) idx of the module will the value on
the x axis and the corresponding feature val in the y axis, and there
will be a seperate line for each channel, and a legend denoting which
line belongs to which channel.

Expected Usage
```
mod_rep_visualizer.generate_plot_visualization(<feature_name>)
```

You can also filter the modules so that only modules with a certain
substring will have their features represented in the plot.

> ** This is intended to be used in a `.ipynb` style notebook**

The tests for this were just visual inspection for two reasons:
    1.) This method does not return anything, it just generates the
    visualization plot
    2.) All the data to create the plot visualization is gotten from
    `generate_filtered_tables` which is already tested, so testing all that
    for this again would be redundant.

Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -435,17 +445,94 @@ def generate_table_visualization(self, feature_filter: str = "", module_fqn_filt
+        x_data: List = []
+        y_data: List[List] = []
+        # the feature will either be a tensor feature or channel feature
+        if tensor_info_features_count == 1:
```
branch condition looks weird, if it can only be one of them, can we have a variable for that, e.g. `is_show_tensor_info` and branch on that

vspenubarthi replied:
it's possible that the user tries to have multiple features included at once for plotting, which is not something that we want and we want to alert them of, which is why we have the explicit looks and checks. I updated it a bit so that it looks less hardcoded.

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -435,17 +445,94 @@ def generate_table_visualization(self, feature_filter: str = "", module_fqn_filt
+            # it should only be single value data, so easy to plot
+            for table_row_num, row in enumerate(tensor_table):
+                row_value = row[ModelReportVisualizer.NUM_NON_FEATURE_TENSOR_HEADERS]
+                if not type(row_value) == str:
```
this also looks a bit hacky, is there a way to make this cleaner?

vspenubarthi replied:
yup added a couple more var declarations to make it more clear what we are using and why we are using it.

vspenubarthi replied:
cleaned it up to be half the size now

jerryzh168 reviewed (approved):
hope we can clean up `_get_plotting_data` a bit more

vspenubarthi committed:
Update on "[ao] Add line plot visualization capability to ModelReportVisualizer"


Summary: This adds the capability to visualize the line plot in the
ModelReportVisualizer. You can visualize line plots of single feature,
and this feature can either be a per-tensor or per-channel feature. If
the feature is per tensor, then the idx of the module is printed as the
x axis and the values of the feature as the y. If the feature is per
channel, then **an** (the first one) idx of the module will the value on
the x axis and the corresponding feature val in the y axis, and there
will be a seperate line for each channel, and a legend denoting which
line belongs to which channel.

Expected Usage
```
mod_rep_visualizer.generate_plot_visualization(<feature_name>)
```

You can also filter the modules so that only modules with a certain
substring will have their features represented in the plot.

> ** This is intended to be used in a `.ipynb` style notebook**

The tests for this were just visual inspection for two reasons:
    1.) This method does not return anything, it just generates the
    visualization plot
    2.) All the data to create the plot visualization is gotten from
    `generate_filtered_tables` which is already tested, so testing all that
    for this again would be redundant.

Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Add line plot visualization capability to ModelReportVisualizer"


Summary: This adds the capability to visualize the line plot in the
ModelReportVisualizer. You can visualize line plots of single feature,
and this feature can either be a per-tensor or per-channel feature. If
the feature is per tensor, then the idx of the module is printed as the
x axis and the values of the feature as the y. If the feature is per
channel, then **an** (the first one) idx of the module will the value on
the x axis and the corresponding feature val in the y axis, and there
will be a seperate line for each channel, and a legend denoting which
line belongs to which channel.

Expected Usage
```
mod_rep_visualizer.generate_plot_visualization(<feature_name>)
```

You can also filter the modules so that only modules with a certain
substring will have their features represented in the plot.

> ** This is intended to be used in a `.ipynb` style notebook**

The tests for this were just visual inspection for two reasons:
    1.) This method does not return anything, it just generates the
    visualization plot
    2.) All the data to create the plot visualization is gotten from
    `generate_filtered_tables` which is already tested, so testing all that
    for this again would be redundant.

Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Add line plot visualization capability to ModelReportVisualizer"


Summary: This adds the capability to visualize the line plot in the
ModelReportVisualizer. You can visualize line plots of single feature,
and this feature can either be a per-tensor or per-channel feature. If
the feature is per tensor, then the idx of the module is printed as the
x axis and the values of the feature as the y. If the feature is per
channel, then **an** (the first one) idx of the module will the value on
the x axis and the corresponding feature val in the y axis, and there
will be a seperate line for each channel, and a legend denoting which
line belongs to which channel.

Expected Usage
```
mod_rep_visualizer.generate_plot_visualization(<feature_name>)
```

You can also filter the modules so that only modules with a certain
substring will have their features represented in the plot.

> ** This is intended to be used in a `.ipynb` style notebook**

The tests for this were just visual inspection for two reasons:
    1.) This method does not return anything, it just generates the
    visualization plot
    2.) All the data to create the plot visualization is gotten from
    `generate_filtered_tables` which is already tested, so testing all that
    for this again would be redundant.

Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
@pytorchbot merge -g
